File: sdk/samples/WFSamples/Technologies/DesignerHosting/VB/WorkflowDesignerReHosting/DesignerShell.vb
'---------------------------------------------------------------------
'  This file is part of the Windows Workflow Foundation SDK Code Samples.
' 
'  Copyright (C) Microsoft Corporation.  All rights reserved.
' 
'This source code is intended only as a supplement to Microsoft
'Development Tools and/or on-line documentation.  See these other
'materials for detailed information regarding Microsoft code samples.
' 
'THIS CODE AND INFORMATION ARE PROVIDED AS IS WITHOUT WARRANTY OF ANY
'KIND, EITHER EXPRESSED OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE
'IMPLIED WARRANTIES OF MERCHANTABILITY AND/OR FITNESS FOR A
'PARTICULAR PURPOSE.
'---------------------------------------------------------------------

Imports System
Imports System.ComponentModel
Imports System.ComponentModel.Design
Imports System.Drawing
Imports System.Windows.Forms
Imports System.Workflow.Activities
Imports System.Workflow.ComponentModel
Imports System.Workflow.ComponentModel.Design
Imports System.Workflow.ComponentModel.Compiler
Imports Microsoft.VisualBasic

Public Class DesignerShell
    Private workflowPanel As WorkflowViewPanel

    Public Sub New()
        ' This call is required by the Windows Form Designer.
        InitializeComponent()

        ' Now initialize the contained components
        Me.workflowPanel = New WorkflowViewPanel(Me)
        Me.workflowPanel.Dock = DockStyle.Fill
        Me.workflowPanel.Name = "WorkflowViewPanel"

        Me.Panel2.Controls.Add(Me.workflowPanel)

        ' Load the default workflow to start with
        Me.workflowPanel.LoadDefaultWorkflow()
    End Sub

    Private Sub zoomNumericUpDown_ValueChanged(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles zoomNumericUpDown.ValueChanged
        If Me.workflowPanel IsNot Nothing Then
            Me.workflowPanel.OnZoomChanged(CInt(Me.zoomNumericUpDown.Value))
        End If
    End Sub

    Private Sub addButton_Click(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles addButton.Click
        Dim viewId As Integer
        Dim rootDesigner As SequentialWorkflowRootDesigner
        rootDesigner = Me.workflowPanel.GetWorkflowView().RootDesigner
        viewId = rootDesigner.ActiveView.ViewId
        If viewId = 1 Then
            Me.workflowPanel.OnCodeActivityAdded()
        Else
            Dim resultBox As DialogResult
            resultBox = MessageBox.Show("This sample supports adding a code activity only in workflow view")
        End If
    End Sub

    Private Sub updateButton_Click(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles updateButton.Click
        Me.workflowPanel.OnCodeActivityUpdated()
    End Sub
End Class

#Region "Class WorkflowViewPanel"

Public NotInheritable Class WorkflowViewPanel
    Inherits Panel
    Implements IServiceProvider

#Region "Members and Constructor"

    Private designSurface As WorkflowDesignSurface
    Private workflowView As WorkflowView
    Private rootActivity As SequentialWorkflowActivity
    Private designerHost As IDesignerHost
    Private parentValue As DesignerShell

    Public Sub New(ByVal parentValue As DesignerShell)
        Me.parentValue = parentValue
    End Sub

#End Region

#Region "Methods"

    Public Function GetWorkflowView() As WorkflowView
        Return Me.workflowView
    End Function

    Protected Overrides Sub Dispose(ByVal disposing As Boolean)
        If disposing Then
            Clear()
        End If
        MyBase.Dispose(disposing)
    End Sub

    Protected Overrides Function GetService(ByVal serviceType As System.Type) As Object
        If Me.designSurface IsNot Nothing Then
            Return Me.designSurface.GetService(serviceType)
        Else
            Return Nothing
        End If
    End Function

    ' Loads precreated workflow. The function is similar to the above function except
    ' instead of creating empty workflow we create workflow with contents

    Friend Sub LoadDefaultWorkflow()
        Clear()

        Me.designSurface = New WorkflowDesignSurface(Me)
        Dim loader As WorkflowLoader = New WorkflowLoader()
        designSurface.BeginLoad(loader)

        designerHost = CType(GetService(GetType(IDesignerHost)), IDesignerHost)
        If Not designerHost Is Nothing Then
            rootActivity = CType(designerHost.CreateComponent(GetType(SequentialWorkflowActivity)), SequentialWorkflowActivity)
            rootActivity.Name = "Service1"

            Dim codeActivity1 As CodeActivity = New CodeActivity()
            rootActivity.Activities.Add(codeActivity1)
            designerHost.RootComponent.Site.Container.Add(codeActivity1)

            Me.workflowView = New WorkflowView(CType(Me.designSurface, IServiceProvider))
            Me.workflowView.AddDesignerMessageFilter(New CustomMessageFilter(CType(Me.designSurface, IServiceProvider), Me.workflowView, loader))

            Controls.Add(Me.workflowView)
            designerHost.Activate()

            Dim selectionService As ISelectionService = CType(GetService(GetType(ISelectionService)), ISelectionService)
            If Not selectionService Is Nothing Then
                AddHandler selectionService.SelectionChanged, New EventHandler(AddressOf OnSelectionChanged)
                Dim selection() As IComponent = New IComponent() {rootActivity}
                selectionService.SetSelectedComponents(selection)
            End If
        End If
    End Sub

    Private Function GetSelectedActivity() As CodeActivity
        Dim selectionService As ISelectionService = CType(GetService(GetType(ISelectionService)), ISelectionService)

        If selectionService IsNot Nothing AndAlso selectionService.SelectionCount <> 0 Then
            Dim selection() As Object = New Object(selectionService.SelectionCount) {}
            selectionService.GetSelectedComponents().CopyTo(selection, 0)

            If selection(0) IsNot Nothing AndAlso TypeOf selection(0) Is IComponent Then
                If TypeOf selection(0) Is CodeActivity Then
                    Return CType(selection(0), CodeActivity)
                End If
            End If
        End If

        Return Nothing
    End Function

    Private Sub OnSelectionChanged(ByVal sender As Object, ByVal e As System.EventArgs)
        parentValue.nameTextBox.Clear()
        parentValue.descriptionTextBox.Clear()
        parentValue.nameTextBox.ReadOnly = True
        parentValue.descriptionTextBox.ReadOnly = True
        parentValue.updateButton.Enabled = False

        Dim codeActivity As CodeActivity = GetSelectedActivity()
        If Not codeActivity Is Nothing Then
            parentValue.nameTextBox.Text = codeActivity.Name
            parentValue.descriptionTextBox.Text = codeActivity.Description
            parentValue.nameTextBox.ReadOnly = False
            parentValue.descriptionTextBox.ReadOnly = False
            parentValue.updateButton.Enabled = True
        End If
    End Sub

    Public Sub OnCodeActivityUpdated()
        Dim codeActivity As CodeActivity = GetSelectedActivity()
        If Not codeActivity Is Nothing Then
            Try
                ' Using PropertyDescriptor to let a designer update the component name
                Dim propertyDescriptor As PropertyDescriptor = TypeDescriptor.GetProperties(codeActivity)("Name")
                propertyDescriptor.SetValue(codeActivity, Me.parentValue.nameTextBox.Text)
                codeActivity.Description = Me.parentValue.descriptionTextBox.Text
            Catch ex As Exception
                MessageBox.Show(Me.Parent, ex.Message, Me.Parent.Text, MessageBoxButtons.OK, MessageBoxIcon.Error)
            End Try
        End If
    End Sub

    Public Function OnCodeActivityAdded() As Boolean
        Dim codeActivity As CodeActivity = New CodeActivity()

        Dim activityPropertiesForm As ActivityProperties = New ActivityProperties()
        activityPropertiesForm.ShowDialog()
        If activityPropertiesForm.Result = DialogResult.OK Then
            For Each activity As Activity In Me.rootActivity.Activities()
                If activity.Name.Equals(activityPropertiesForm.ActivityName, StringComparison.OrdinalIgnoreCase) Then
                    MessageBox.Show(Me.Parent, "Cannot add new CodeActivity. The CodeActivity with name '" + activityPropertiesForm.ActivityName + "' already exists.", Me.Parent.Text, MessageBoxButtons.OK, MessageBoxIcon.Error)
                    Return False
                End If
            Next

            codeActivity.Name = activityPropertiesForm.ActivityName
            codeActivity.Description = activityPropertiesForm.Description

            Me.rootActivity.Activities.Add(codeActivity)
            Me.designerHost.RootComponent.Site.Container.Add(codeActivity)
            Me.workflowView.Update()
        End If

        Return True
    End Function

    Public Sub OnZoomChanged(ByVal zoomFactor As Integer)
        Me.workflowView.Zoom = zoomFactor
        Me.workflowView.Update()
    End Sub

    Private Sub Clear()
        If Me.designSurface Is Nothing Then
            Return
        Else
            Me.designSurface.Dispose()
        End If

        If Me.workflowView IsNot Nothing AndAlso Controls.Contains(Me.workflowView) Then
            Controls.Remove(Me.workflowView)
        End If
        Me.workflowView = Nothing
    End Sub

    Protected Overrides Sub OnLayout(ByVal levent As LayoutEventArgs)
        MyBase.OnLayout(levent)
        If Me.workflowView IsNot Nothing AndAlso Not Me.workflowView.IsDisposed Then
            Me.workflowView.Width = Me.Width
            Me.workflowView.Height = Me.Height
        End If
    End Sub

    Public Sub InvokeStandardCommand(ByVal cmd As CommandID)
        Try
            Dim menuService As IMenuCommandService = CType(GetService(GetType(IMenuCommandService)), IMenuCommandService)
            If Not menuService Is Nothing Then
                menuService.GlobalInvoke(cmd)
            End If
        Catch
            'We eat exceptions as some of the operations are not supported in samples
        End Try
    End Sub
#End Region

#Region "IServiceProvider Members"

    Private Function IServiceProvider_GetService(ByVal serviceType As System.Type) As Object Implements System.IServiceProvider.GetService
        Return GetService(serviceType)
    End Function

#End Region

End Class

#End Region

#Region "Class WorkflowDesignSurface"

' Design Surface is used to provide services.
Friend NotInheritable Class WorkflowDesignSurface
    Inherits DesignSurface

    Friend Sub New(ByVal serviceProvider As IServiceProvider)
        Me.ServiceContainer.AddService(GetType(IMenuCommandService), New MenuCommandService(Me.ServiceContainer))

        Dim typeProvider As TypeProvider = New TypeProvider(serviceProvider)
        typeProvider.AddAssemblyReference(GetType(String).Assembly.Location)
        Me.ServiceContainer.AddService(GetType(ITypeProvider), typeProvider, True)
    End Sub
End Class

#End Region

#Region "Class CustomMessageFilter"
' All Coordinates passed in physical coordinate system
' Some of the functions will have coordinates in screen coordinates ie ShowContextMenu
Friend NotInheritable Class CustomMessageFilter
    Inherits WorkflowDesignerMessageFilter

#Region "Members and Constructor"

    Private mouseDown As Boolean
    Private serviceProvider As IServiceProvider
    Private workflowView As WorkflowView
    Private loader As WorkflowDesignerLoader

    Public Sub New(ByVal provider As IServiceProvider, ByVal workflowView As WorkflowView, ByVal loader As WorkflowDesignerLoader)
        Me.serviceProvider = provider
        Me.workflowView = workflowView
        Me.loader = loader
    End Sub

#End Region

#Region "WorkflowDesignerMessageFilter Overridables"

    Protected Overrides Function OnMouseDown(ByVal eventArgs As System.Windows.Forms.MouseEventArgs) As Boolean
        ' Allow other components to process this event by not returning true.
        mouseDown = True
        Return False
    End Function

    Protected Overrides Function OnMouseMove(ByVal eventArgs As System.Windows.Forms.MouseEventArgs) As Boolean
        ' Allow other components to process this event by not returning true.
        If mouseDown Then
            workflowView.ScrollPosition = New Point(eventArgs.X, eventArgs.Y)
        End If
        Return False
    End Function

    Protected Overrides Function OnMouseUp(ByVal eventArgs As MouseEventArgs) As Boolean
        ' Allow other components to process this event by not returning true.
        mouseDown = False
        Return False
    End Function

    Protected Overrides Function OnMouseDoubleClick(ByVal eventArgs As MouseEventArgs) As Boolean
        mouseDown = False
        Return True
    End Function

    Protected Overrides Function OnMouseEnter(ByVal eventArgs As MouseEventArgs) As Boolean
        ' Allow other components to process this event by not returning true.
        mouseDown = False
        Return False
    End Function

    Protected Overrides Function OnMouseHover(ByVal eventArgs As MouseEventArgs) As Boolean
        ' Allow other components to process this event by not returning true.
        mouseDown = False
        Return False
    End Function

    Protected Overrides Function OnMouseLeave() As Boolean
        ' Allow other components to process this event by not returning true.
        mouseDown = False
        Return False
    End Function

    Protected Overrides Function OnMouseWheel(ByVal eventArgs As MouseEventArgs) As Boolean
        mouseDown = False
        Return True
    End Function

    Protected Overrides Function OnMouseCaptureChanged() As Boolean
        ' Allow other components to process this event by not returning true.
        mouseDown = False
        Return False
    End Function

    Protected Overrides Function OnDragEnter(ByVal eventArgs As DragEventArgs) As Boolean
        Return True
    End Function

    Protected Overrides Function OnDragOver(ByVal eventArgs As DragEventArgs) As Boolean
        Return True
    End Function

    Protected Overrides Function OnDragLeave() As Boolean
        Return True
    End Function

    Protected Overrides Function OnDragDrop(ByVal eventArgs As DragEventArgs) As Boolean
        Return True
    End Function

    Protected Overrides Function OnGiveFeedback(ByVal gfbevent As GiveFeedbackEventArgs) As Boolean
        Return True
    End Function

    Protected Overrides Function OnQueryContinueDrag(ByVal qcdevent As QueryContinueDragEventArgs) As Boolean
        Return True
    End Function

    Protected Overrides Function OnKeyDown(ByVal eventArgs As KeyEventArgs) As Boolean
        If eventArgs.KeyCode = Keys.Delete Then
            Dim selectionService As ISelectionService = CType(serviceProvider.GetService(GetType(ISelectionService)), ISelectionService)
            If selectionService IsNot Nothing AndAlso TypeOf selectionService.PrimarySelection Is CodeActivity Then
                Dim codeActivityComponent As CodeActivity = CType(selectionService.PrimarySelection, CodeActivity)
                Dim parentActivity As CompositeActivity = codeActivityComponent.Parent
                If parentActivity IsNot Nothing Then
                    parentActivity.Activities.Remove(codeActivityComponent)
                    Me.ParentView.Update()
                End If
                loader.RemoveActivityFromDesigner(codeActivityComponent)
            End If
        End If
        Return True
    End Function

    Protected Overrides Function OnKeyUp(ByVal eventArgs As KeyEventArgs) As Boolean
        Return True
    End Function

    Protected Overrides Function OnShowContextMenu(ByVal menuPoint As Point) As Boolean
        Return True
    End Function

#End Region

End Class

#End Region

File: sdk/samples/WFSamples/Technologies/DesignerHosting/CS/WorkflowDesignerReHosting/DesignerShell.cs
//---------------------------------------------------------------------
//  This file is part of the Windows Workflow Foundation SDK Code Samples.
// 
//  Copyright (C) Microsoft Corporation.  All rights reserved.
// 
//This source code is intended only as a supplement to Microsoft
//Development Tools and/or on-line documentation.  See these other
//materials for detailed information regarding Microsoft code samples.
// 
//THIS CODE AND INFORMATION ARE PROVIDED AS IS WITHOUT WARRANTY OF ANY
//KIND, EITHER EXPRESSED OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE
//IMPLIED WARRANTIES OF MERCHANTABILITY AND/OR FITNESS FOR A
//PARTICULAR PURPOSE.
//---------------------------------------------------------------------

using System;
using System.ComponentModel;
using System.ComponentModel.Design;
using System.Drawing;
using System.Windows.Forms;
using System.Workflow.Activities;
using System.Workflow.ComponentModel;
using System.Workflow.ComponentModel.Design;
using System.Workflow.ComponentModel.Compiler;

namespace Microsoft.Samples.Workflow.WorkflowDesignerRehosting
{
    #region Class DesignerShell

    public partial class DesignerShell : Form
    {
        #region Members and Constructors

        private WorkflowViewPanel workflowPanel;

        public DesignerShell()
        {
            InitializeComponent();

            //Now initialize the contained components
            this.workflowPanel = new WorkflowViewPanel(this);
            this.workflowPanel.Dock = DockStyle.Fill;
            this.workflowPanel.Name = "WorkflowViewPanel";

            this.Panel2.Controls.Add(this.workflowPanel);

            //load the default workflow to start with
            this.workflowPanel.LoadDefaultWorkflow();
        }

        #endregion

        private void zoomNumericUpDown_ValueChanged(object sender, EventArgs e)
        {
            if (this.workflowPanel != null)
                this.workflowPanel.OnZoomChanged(Convert.ToInt16(this.zoomNumericUpDown.Value));
        }

        private void addButton_Click(object sender, EventArgs e)
        {
            SequentialWorkflowRootDesigner rootDesigner = this.workflowPanel.GetWorkflowView().RootDesigner as SequentialWorkflowRootDesigner;
            int viewId = rootDesigner.ActiveView.ViewId;
            if(viewId == 1)
            {
               this.workflowPanel.OnCodeActivityAdded();
            }
            else
            {
               DialogResult resultBox = MessageBox.Show("This sample supports adding a code activity only in workflow view");
            }            
        }

        private void updateButton_Click(object sender, EventArgs e)
        {
            this.workflowPanel.OnCodeActivityUpdated();
        }
    }

    #endregion

    #region Class WorkflowViewPanel

    public sealed class WorkflowViewPanel : Panel, IServiceProvider
    {
        #region Members and Constructor

        private WorkflowDesignSurface designSurface;
        private WorkflowView workflowView;
        private SequentialWorkflowActivity rootActivity;
        private IDesignerHost designerHost;
        private DesignerShell parent;

        public WorkflowViewPanel(DesignerShell parent)
        {
            this.parent = parent;
        }

        #endregion

        #region Properties and Methods

        public WorkflowView GetWorkflowView()
        {
            return this.workflowView;
        }

        protected override void Dispose(bool disposing)
        {
            if (disposing)
                Clear();
            base.Dispose(disposing);
        }
        
        protected override object GetService(Type serviceType)
        {
            if (this.designSurface != null)
                return this.designSurface.GetService(serviceType);
            else
                return null;
        }

        // Loads precreated workflow. The function is similar to the above function except
        // instead of creating an empty workflow we create workflow with contents
        internal void LoadDefaultWorkflow()
        {
            Clear();

            this.designSurface = new WorkflowDesignSurface(this);
            WorkflowLoader loader = new WorkflowLoader();
            this.designSurface.BeginLoad(loader);
            
            this.designerHost = GetService(typeof(IDesignerHost)) as IDesignerHost;
            if (this.designerHost != null)
            {
                this.rootActivity = (SequentialWorkflowActivity)this.designerHost.CreateComponent(typeof(SequentialWorkflowActivity));
                this.rootActivity.Name = "Service1";

                CodeActivity codeActivity1 = new CodeActivity();
                this.rootActivity.Activities.Add(codeActivity1);
                this.designerHost.RootComponent.Site.Container.Add(codeActivity1);

                this.workflowView = new WorkflowView(this.designSurface as IServiceProvider);
                this.workflowView.AddDesignerMessageFilter(new CustomMessageFilter(this.designSurface as IServiceProvider, this.workflowView, loader));

                Controls.Add(this.workflowView);
                this.designerHost.Activate();

                ISelectionService selectionService = (ISelectionService)GetService(typeof(ISelectionService));
                if (selectionService != null)
                {
                    selectionService.SelectionChanged += new EventHandler(OnSelectionChanged);
                    IComponent[] selection = new IComponent[] { rootActivity };
                    selectionService.SetSelectedComponents(selection);
                }
            }
        }

        private CodeActivity GetSelectedActivity()
        {
            ISelectionService selectionService = (ISelectionService)GetService(typeof(ISelectionService));

            if ((selectionService != null) && (selectionService.SelectionCount != 0))
            {
                object[] selection = new object[selectionService.SelectionCount];
                selectionService.GetSelectedComponents().CopyTo(selection, 0);

                if ((selection[0] != null) && (selection[0] is IComponent))
                {
                    if (selection[0] is CodeActivity)
                    {
                        return (CodeActivity)selection[0];
                    }
                }
            }

            return null;
        }

        private void OnSelectionChanged(object sender, System.EventArgs e)
        {
            this.parent.nameTextBox.Clear();
            this.parent.descriptionTextBox.Clear();
            this.parent.nameTextBox.ReadOnly = this.parent.descriptionTextBox.ReadOnly = true;
            this.parent.updateButton.Enabled = false;

            CodeActivity codeActivity = GetSelectedActivity();
            if (codeActivity != null)
            {
                this.parent.nameTextBox.Text = codeActivity.Name;
                this.parent.descriptionTextBox.Text = codeActivity.Description;
                this.parent.nameTextBox.ReadOnly = this.parent.descriptionTextBox.ReadOnly = false;
                this.parent.updateButton.Enabled = true;
            }
        }

        public void OnCodeActivityUpdated()
        {
            CodeActivity codeActivity = GetSelectedActivity();
            if (codeActivity != null)
            {
                try
                {
                    // Using PropertyDescriptor to let a designer update the component name
                    PropertyDescriptor propertyDescriptor = TypeDescriptor.GetProperties(codeActivity)["Name"];
                    propertyDescriptor.SetValue(codeActivity, this.parent.nameTextBox.Text);
                    codeActivity.Description = this.parent.descriptionTextBox.Text;
                }
                catch (Exception ex)
                {
                    MessageBox.Show(this.parent, ex.Message, this.parent.Text, MessageBoxButtons.OK, MessageBoxIcon.Error);
                }
            }
        }

        public bool OnCodeActivityAdded()
        {

            CodeActivity codeActivity = new CodeActivity();

            ActivityProperties activityPropertiesForm = new ActivityProperties();
            activityPropertiesForm.ShowDialog();
            if (activityPropertiesForm.Result == DialogResult.OK)
            {
                foreach (Activity activity in this.rootActivity.Activities)
                {
                    if (activity.Name.Equals(activityPropertiesForm.ActivityName, StringComparison.OrdinalIgnoreCase))
                    {
                        MessageBox.Show(this.parent, "Cannot add new CodeActivity. The CodeActivity with name '" + activityPropertiesForm.ActivityName + "' already exists.", this.parent.Text, MessageBoxButtons.OK, MessageBoxIcon.Error);
                        return false;
                    }
                }

                codeActivity.Name = activityPropertiesForm.ActivityName;
                codeActivity.Description = activityPropertiesForm.Description;

                this.rootActivity.Activities.Add(codeActivity);
                this.designerHost.RootComponent.Site.Container.Add(codeActivity);
                this.workflowView.Update();
            }

            return true;
        }

        public void OnZoomChanged(int zoomFactor)
        {
            this.workflowView.Zoom = zoomFactor;
            this.workflowView.Update();
        }

        private void Clear()
        {
            if (this.designSurface == null)
                return;
            else
                this.designSurface.Dispose();

            if (this.workflowView != null && Controls.Contains(this.workflowView))
                Controls.Remove(this.workflowView);
            this.workflowView = null;
        }

        protected override void OnLayout(LayoutEventArgs levent)
        {
            base.OnLayout(levent);
            if (this.workflowView != null && !this.workflowView.IsDisposed)
            {
                this.workflowView.Width = this.Width;
                this.workflowView.Height = this.Height;
            }
        }

        public void InvokeStandardCommand(CommandID cmd)
        {
            try
            {
                IMenuCommandService menuService = GetService(typeof(IMenuCommandService)) as IMenuCommandService;
                if (menuService != null)
                    menuService.GlobalInvoke(cmd);
            }
            catch
            {
                //We eat exceptions as some of the operations are not supported in samples
            }
        }

        #endregion

        #region IServiceProvider Members

        object System.IServiceProvider.GetService(Type serviceType)
        {
            return GetService(serviceType);
        }

        #endregion
    }
    #endregion

    #region Class WorkflowDesignSurface

    // Design Surface is used to provide services.
    internal sealed class WorkflowDesignSurface : DesignSurface
    {
        internal WorkflowDesignSurface(IServiceProvider serviceProvider)
        {
            this.ServiceContainer.AddService(typeof(IMenuCommandService), new MenuCommandService(this.ServiceContainer));

            TypeProvider typeProvider = new TypeProvider(serviceProvider);
            typeProvider.AddAssemblyReference(typeof(string).Assembly.Location);
            this.ServiceContainer.AddService(typeof(ITypeProvider), typeProvider, true);
        }
    }

    #endregion

    #region Class CustomMessageFilter

    //All Coordinates passed in physical coordinate system
    //Some of the functions will have coordinates in screen coordinates, that is, ShowContextMenu.
    internal sealed class CustomMessageFilter : WorkflowDesignerMessageFilter
    {
        #region Members and Constructor

        private bool mouseDown;
        private IServiceProvider serviceProvider;
        private WorkflowView workflowView;
        private WorkflowDesignerLoader loader;

        public CustomMessageFilter(IServiceProvider provider, WorkflowView workflowView, WorkflowDesignerLoader loader)
        {
            this.serviceProvider = provider;
            this.workflowView = workflowView;
            this.loader = loader;
        }

        #endregion

        #region MessageFilter Overridables

        protected override bool OnMouseDown(MouseEventArgs eventArgs)
        {
            //Allow other components to process this event by not returning true.
            this.mouseDown = true;
            return false;
        }

        protected override bool OnMouseMove(MouseEventArgs eventArgs)
        {
            //Allow other components to process this event by not returning true.
            if (mouseDown)
            {
                workflowView.ScrollPosition = new Point(eventArgs.X, eventArgs.Y);
            }
            return false;
        }

        protected override bool OnMouseUp(MouseEventArgs eventArgs)
        {
            //Allow other components to process this event by not returning true.
            mouseDown = false;
            return false;
        }

        protected override bool OnMouseDoubleClick(MouseEventArgs eventArgs)
        {
            mouseDown = false;
            return true;
        }

        protected override bool OnMouseEnter(MouseEventArgs eventArgs)
        {
            //Allow other components to process this event by not returning true.
            mouseDown = false;
            return false;
        }

        protected override bool OnMouseHover(MouseEventArgs eventArgs)
        {
            //Allow other components to process this event by not returning true.
            mouseDown = false;
            return false;
        }

        protected override bool OnMouseLeave()
        {
            //Allow other components to process this event by not returning true.
            mouseDown = false;
            return false;
        }

        protected override bool OnMouseWheel(MouseEventArgs eventArgs)
        {
            mouseDown = false;
            return true;
        }

        protected override bool OnMouseCaptureChanged()
        {
            //Allow other components to process this event by not returning true.
            mouseDown = false;
            return false;
        }

        protected override bool OnDragEnter(DragEventArgs eventArgs)
        {
            return true;
        }

        protected override bool OnDragOver(DragEventArgs eventArgs)
        {
            return true;
        }

        protected override bool OnDragLeave()
        {
            return true;
        }

        protected override bool OnDragDrop(DragEventArgs eventArgs)
        {
            return true;
        }

        protected override bool OnGiveFeedback(GiveFeedbackEventArgs gfbevent)
        {
            return true;
        }

        protected override bool OnQueryContinueDrag(QueryContinueDragEventArgs qcdevent)
        {
            return true;
        }

        protected override bool OnKeyDown(KeyEventArgs eventArgs)
        {
            if (eventArgs.KeyCode == Keys.Delete)
            {
                ISelectionService selectionService = (ISelectionService)serviceProvider.GetService(typeof(ISelectionService));
                if (selectionService != null && selectionService.PrimarySelection is CodeActivity)
                {
                    CodeActivity codeActivityComponent = (CodeActivity)selectionService.PrimarySelection;
                    CompositeActivity parentActivity = codeActivityComponent.Parent;
                    if (parentActivity != null)
                    {
                        parentActivity.Activities.Remove(codeActivityComponent);
                        this.ParentView.Update();
                    }
                    loader.RemoveActivityFromDesigner(codeActivityComponent);
                    
                }
            }
            return true;
        }

        protected override bool OnKeyUp(KeyEventArgs eventArgs)
        {
            return true;
        }

        protected override bool OnShowContextMenu(Point menuPoint)
        {
            return true;
        }

        #endregion
    }

    #endregion
}